comit/src/expiries.rs

```diff
@@ -0,0 +1,488 @@
+    where
+        C: CurrentTime,
+    {
+        if current_state == BobState::Done {
+            return BobAction::NoFurtherAction;
+        }
+
+        let funded = current_state.includes_fund();
+
+        if self.alpha_expiry_has_elapsed() {
+            if funded {
+                return BobAction::Refund;
+            }
+            return BobAction::Abort;
+        };
+
+        if !self.bob_has_time_to_complete(current_state) {
+            if funded {
+                return BobAction::WaitForRefund;
+            }
+            return BobAction::Abort;
+        }
+
+        let (next_action, _state) = current_state.next();
+        next_action
+    }
+
+    fn alpha_expiry_has_elapsed(&self) -> bool
+    where
+        C: CurrentTime,
+    {
+        let now = self.connector.alpha_current_time();
+        now > self.alpha_expiry
+    }
+
+    fn beta_expiry_has_elapsed(&self) -> bool
+    where
+        C: CurrentTime,
+    {
+        let now = self.connector.beta_current_time();
+        now > self.beta_expiry
+    }
+}
+
+/// The minimum time we should allow for Alice to transition from
+/// `current_state` to done.
+fn period_for_alice_to_complete(config: &Config, current_state: AliceState) -> Duration {
+    // Tail call recursion with an accumulator.
+    fn period_to_complete(config: &Config, state: AliceState, acc: Duration) -> Duration {
+        if state == AliceState::Done {
+            return acc;
+        }
+
+        let (_action, next_state) = state.next();
+        let transition_period = state.transition_period();
+
+        period_to_complete(config, next_state, acc + transition_period)
+    }
+
+    period_to_complete(config, current_state, Duration::zero())
+}
+
+/// The minimum time we should allow for Bob to transition from
+/// `current_state` to done.
+fn period_for_bob_to_complete(config: &Config, current_state: BobState) -> Duration {
+    // Tail call recursion with an accumulator.
+    fn period_to_complete(config: &Config, state: BobState, acc: Duration) -> Duration {
+        if state == BobState::Done {
+            return acc;
+        }
+
+        let (_action, next_state) = state.next();
+        let transition_period = state.transition_period();
+
+        period_to_complete(config, next_state, acc + transition_period)
+    }
+
+    period_to_complete(config, current_state, Duration::zero())
+}
+
+/// Happy path states for Alice.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum AliceState {
+    // FIXME: Do we need a pre-created state in order to give Alice one more TTA period
+    // before `created`?
+    Created,
+    FundAlphaTransactionSent,
+    AlphaFunded, // Implies alpha fund transaction confirmed configured number of times.
+    BetaFunded,  // Implies beta fund transaction confirmed configured number of times.
+    RedeemBetaTransactionSent,
+    Done, // Implies beta redeem transaction confirmed configured number of times.
```

Comment:
I reckon these comments should go as rust-docs on the variants :)

Reply:
Can do.